<a href="https://colab.research.google.com/github/congtuan99/learning-data/blob/main/day1_penguins.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Partie 1 : Initiation à Pandas

In [1]:
import pandas as pd
import seaborn as sns

In [2]:
df = sns.load_dataset("penguins")

In [3]:
df.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


In [4]:
df.shape

(344, 7)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [6]:
df.describe()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


In [7]:
df["species"]

,species
0,Adelie
1,Adelie
2,Adelie
3,Adelie
4,Adelie
...,...
339,Gentoo
340,Gentoo
341,Gentoo
342,Gentoo


In [8]:
df["species"].value_counts()

,count
species,
Adelie,152
Gentoo,124
Chinstrap,68


In [9]:
df[(df["species"]=="Gentoo")&(df["body_mass_g"] > 4500)]

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
221,Gentoo,Biscoe,50.0,16.3,230.0,5700.0,Male
223,Gentoo,Biscoe,50.0,15.2,218.0,5700.0,Male
224,Gentoo,Biscoe,47.6,14.5,215.0,5400.0,Male
225,Gentoo,Biscoe,46.5,13.5,210.0,4550.0,Female
226,Gentoo,Biscoe,45.4,14.6,211.0,4800.0,Female
...,...,...,...,...,...,...,...
338,Gentoo,Biscoe,47.2,13.7,214.0,4925.0,Female
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0,Female
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0,Male
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0,Female


In [10]:

df.iloc[5]
df.iloc[0:7]
df.loc[3]

,3
species,Adelie
island,Torgersen
bill_length_mm,NaN
bill_depth_mm,NaN
flipper_length_mm,NaN
body_mass_g,NaN
sex,NaN


In [11]:
df[df["species"]=="Adelie"]["body_mass_g"].mean()


np.float64(3700.662251655629)

# Partie 2: SQL avec duckdb

In [12]:
!pip install duckdb -q
import duckdb

In [13]:
duckdb.sql("select * from df limit 5")

┌─────────┬───────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┐
│ species │  island   │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │
│ varchar │  varchar  │     double     │    double     │      double       │   double    │ varchar │
├─────────┼───────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┤
│ Adelie  │ Torgersen │           39.1 │          18.7 │             181.0 │      3750.0 │ Male    │
│ Adelie  │ Torgersen │           39.5 │          17.4 │             186.0 │      3800.0 │ Female  │
│ Adelie  │ Torgersen │           40.3 │          18.0 │             195.0 │      3250.0 │ Female  │
│ Adelie  │ Torgersen │           NULL │          NULL │              NULL │        NULL │ NULL    │
│ Adelie  │ Torgersen │           36.7 │          19.3 │             193.0 │      3450.0 │ Female  │
└─────────┴───────────┴────────────────┴───────────────┴───────────────────┴─────────────┴─

In [14]:
duckdb.sql("select species, body_mass_g from df")

┌─────────┬─────────────┐
│ species │ body_mass_g │
│ varchar │   double    │
├─────────┼─────────────┤
│ Adelie  │      3750.0 │
│ Adelie  │      3800.0 │
│ Adelie  │      3250.0 │
│ Adelie  │        NULL │
│ Adelie  │      3450.0 │
│ Adelie  │      3650.0 │
│ Adelie  │      3625.0 │
│ Adelie  │      4675.0 │
│ Adelie  │      3475.0 │
│ Adelie  │      4250.0 │
│   ·     │         ·   │
│   ·     │         ·   │
│   ·     │         ·   │
│ Gentoo  │      4375.0 │
│ Gentoo  │      5850.0 │
│ Gentoo  │      4875.0 │
│ Gentoo  │      6000.0 │
│ Gentoo  │      4925.0 │
│ Gentoo  │        NULL │
│ Gentoo  │      4850.0 │
│ Gentoo  │      5750.0 │
│ Gentoo  │      5200.0 │
│ Gentoo  │      5400.0 │
├─────────┴─────────────┤
│  344 rows (20 shown)  │
└───────────────────────┘

In [15]:
duckdb.sql("select * from df where species='Gentoo' and body_mass_g>4500")


┌─────────┬─────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┐
│ species │ island  │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │
│ varchar │ varchar │     double     │    double     │      double       │   double    │ varchar │
├─────────┼─────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┤
│ Gentoo  │ Biscoe  │           50.0 │          16.3 │             230.0 │      5700.0 │ Male    │
│ Gentoo  │ Biscoe  │           50.0 │          15.2 │             218.0 │      5700.0 │ Male    │
│ Gentoo  │ Biscoe  │           47.6 │          14.5 │             215.0 │      5400.0 │ Male    │
│ Gentoo  │ Biscoe  │           46.5 │          13.5 │             210.0 │      4550.0 │ Female  │
│ Gentoo  │ Biscoe  │           45.4 │          14.6 │             211.0 │      4800.0 │ Female  │
│ Gentoo  │ Biscoe  │           46.7 │          15.3 │             219.0 │      5200.0 │ Male    │
│ Gentoo  

In [16]:
duckdb.sql("select * from df order by body_mass_g DESC limit 3")

┌─────────┬─────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┐
│ species │ island  │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │
│ varchar │ varchar │     double     │    double     │      double       │   double    │ varchar │
├─────────┼─────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┤
│ Gentoo  │ Biscoe  │           49.2 │          15.2 │             221.0 │      6300.0 │ Male    │
│ Gentoo  │ Biscoe  │           59.6 │          17.0 │             230.0 │      6050.0 │ Male    │
│ Gentoo  │ Biscoe  │           51.1 │          16.3 │             220.0 │      6000.0 │ Male    │
└─────────┴─────────┴────────────────┴───────────────┴───────────────────┴─────────────┴─────────┘

# Partie 3: Analyse - Comparaison des trois espèces

In [17]:
# Comparer les trois espèces
# 1: la masse corporelle moyenne
x1 = df[df["species"]=="Adelie"]["body_mass_g"].mean()
x2 = df[df["species"]=="Gentoo"]["body_mass_g"].mean()
x3 = df[df["species"]=="Chinstrap"]["body_mass_g"].mean()
print(x1,x2,x3)

3700.662251655629 5076.016260162602 3733.0882352941176


In [18]:
# 2: la longueur des nageoires moyenne
y1 = df[df["species"]=="Adelie"]["flipper_length_mm"].mean()
y2 = df[df["species"]=="Gentoo"]["flipper_length_mm"].mean()
y3 = df[df["species"]=="Chinstrap"]["flipper_length_mm"].mean()
print(y1,y2,y3)

189.95364238410596 217.1869918699187 195.8235294117647


## Commentaire:
On compare d'abord la masse corporelle moyenne des espèces. L'espèce Gentoo a une masse nettement supérieure à celle des deux autres, soit environ 5076 g, contre environ 3700 g pour les Adelie et 3733 g pour les Chinstrap.
Ensuite, on compare la longueur des nageoires. La même tendance se confirme : les Gentoo ont les nageoires les plus longues, environ 217 mm, tandis que les Adelie et les Chinstrap ont une longueur à peu près équivalente, environ 190 mm et 196 mm.
En résumé, les Gentoo se distinguent nettement des autres par leur grande taille, aussi bien en masse corporelle qu'en longueur des nageoires, alors que les Chinstrap sont légèrement plus grands que les Adelie.